tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp

```diff
@@ -1093,137 +1090,105 @@ bool Trajectory::isValid() const { return valid_; }
 
 double Trajectory::getDuration() const { return trajectory_.back().time_; }
 
-bool Trajectory::assignData(TrajectoryContainer& trajectory, double path_tolerance) const
+bool Trajectory::assignData(TrajectoryContainer& trajectory) const
 {
-  auto it = trajectory_.begin();
+  Eigen::VectorXd path_lengths = Eigen::VectorXd::Zero(trajectory.size());
+  double length = 0;
+  for (Eigen::Index i = 1; i < trajectory.size(); ++i)
+  {
+    length += (trajectory.getPosition(i) - trajectory.getPosition(i - 1)).norm();
+    path_lengths[i] = length;
+  }
+
+  if (tesseract_common::almostEqualRelativeAndAbs(length, 0))
+    throw std::runtime_error("Tried to assign trajectory with length zero.");
+
+  double scale = path_.getLength() / length;
+  path_lengths = scale * path_lengths;
+
   long cnt = 0;
-  double error = 0;
-  double prev_error = 1;
-  bool hit_tolerance = false;
-  bool error_increasing = false;
+  Eigen::Index sample_size = std::max(static_cast<Eigen::Index>(getDuration() / 0.01), trajectory.size());
+  Eigen::VectorXd times = Eigen::VectorXd::LinSpaced(sample_size, 0, getDuration());
   for (long i = 0; i < trajectory.size(); ++i)
   {
-    const Eigen::VectorXd& p = trajectory.getPosition(i);
+    Eigen::VectorXd uv;
+    Eigen::VectorXd ua;
+    double time{ 0 };
 
     if (i == 0)
     {
-      // Set first point
-      Eigen::VectorXd uv = getVelocity(trajectory_.front().time_).head(trajectory.dof());
-      Eigen::VectorXd ua = getAcceleration(trajectory_.front().time_).head(trajectory.dof());
-      trajectory.setData(0, uv, ua, it->time_);
-      ++cnt;
+      uv = getVelocity(0).head(trajectory.dof());
+      ua = getAcceleration(0).head(trajectory.dof());
+      time = 0;
+      cnt++;
     }
-    else if (i == (trajectory.size() - 1))
+    else if (i == trajectory.size() - 1)
     {
-      // Set last point
-      Eigen::VectorXd uv = getVelocity(trajectory_.back().time_).head(trajectory.dof());
-      Eigen::VectorXd ua = getAcceleration(trajectory_.back().time_).head(trajectory.dof());
-      trajectory.setData(trajectory.size() - 1, uv, ua, it->time_);
-      ++cnt;
+      time = getDuration();
+      uv = getVelocity(time).head(trajectory.dof());
+      ua = getAcceleration(time).head(trajectory.dof());
+      cnt++;
     }
     else
     {
-      for (; it != trajectory_.end(); ++it)
+      const Eigen::VectorXd& p = trajectory.getPosition(i);
+      double path_length = path_lengths[i];
+      // calculate distance from this point to all other points in the trajectory
+
+      Eigen::VectorXd x(sample_size);
+      Eigen::VectorXd dists(sample_size);
+      for (Eigen::Index j = 0; j < sample_size; ++j)
       {
-        Eigen::VectorXd compare_p = getPosition(it->time_).head(trajectory.dof());
-        error = (p - compare_p).norm();
+        PathData data = getPathData(times[j]);
+        x[j] = data.path_pos;
+        dists[j] = (getPosition(times[j]).head(trajectory.dof()) - p).norm();
+      }
```

Comment:
It might be nice to put some blocks like this into functions with descriptive names that we can test later.

Reply:
I agree. I looked quickly and it will be more work to break this out into a separate function than time I have at the moment.